### PR TITLE
Activity stream loads previous activities of a room

### PR DESCRIPTION
**Pull request by lalli-flores**
Tested with "Hellen Keller" data:
https://github.com/lalli-flores/components/tree/hellen-keller

I would prefer to have this tested with different data set.

### CODE REVIEW

File: .storybook/setupTests.js [hunk 1]
```diff
@@ -2,6 +2,8 @@ import React from 'react';
 import Enzyme, {shallow, render, mount} from 'enzyme';
 import Adapter from 'enzyme-adapter-react-16';
 import * as rxjs from 'rxjs';
+import 'core-js/stable';
```
Comment (akoushke) [inline .storybook/setupTests.js]: what are these for?

File: src/components/hooks/useActivityStream.js [hunk 2]
```diff
@@ -1,53 +1,129 @@
...
+
+  const handleScroll = (event) => {
+    // Only load more data when user has reached the top
+    const isViewportTop = event.target.scrollTop === 0;
+
+    if (!showLoader && isViewportTop && roomsAdapter.hasMoreActivities(roomID)) {
+      setShowLoader(true);
+
+      // Throttle scroll
+      setTimeout(async () => {
```
Comment (akoushke) [inline src/components/hooks/useActivityStream.js]: is the `setTimeout` still valid here? I don't think we would need it anymore, since we are awaiting.

File: src/components/hooks/useActivityStream.js [hunk 3]
```diff
@@ -1,53 +1,129 @@
...
+        loadPreviousActivities();
+        // if there are no more activities to load, scroll to last activity
+      } else {
+        lastElementRef.current.scrollIntoView();
+        setScrollToLastActivity(false);
+      }
+    } else {
+      // Scroll to bottom once after loading enough content
+      // eslint-disable-next-line no-lonely-if
+      if (scrollToLastActivity) {
```
Comment (akoushke) [inline src/components/hooks/useActivityStream.js]: why not `else if` ?

File: src/data/activities.json [hunk 4]
```diff
@@ -62,5 +62,181 @@
     "personID": "default",
     "created": "old",
     "displayHeader": true
+  },
+  "hellen-keller-1": {
```
Comment (akoushke) [inline src/data/activities.json]: great, more data (y)

File: src/components/hooks/useActivityStream.js [hunk 5]
```diff
@@ -1,53 +1,129 @@
...
+        lastElementRef.current.scrollIntoView();
+        setScrollToLastActivity(false);
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [activities]);
+
+  // Attach scroll event listener and subscribe to future updates on load
+  useEffect(() => {
```
Comment (akoushke) [inline src/components/hooks/useActivityStream.js]: so two `useEffect`? elaborate a bit

File: src/components/WebexActivityStream/WebexActivityStream.scss [hunk 6]
```diff
@@ -1,3 +1,71 @@
+// For Storybook, so that activity stream loads content properly
+#root {
```
Comment (akoushke) [inline src/components/WebexActivityStream/WebexActivityStream.scss]: why don't you put this at the top level? this might get lost here. maybe `scss` file to be imported into setupFile

File: .storybook/setupTests.js (hunk 1 above)
Comment (akoushke) [reply .storybook/setupTests.js]: maybe a short comment.

File: src/adapters/RoomsAdapter.js [hunk 7]
```diff
@@ -47,26 +63,42 @@ export default class RoomsAdapter extends WebexAdapter {
...
+   * Returns [] if `hasMoreActivities` returns false, otherwise it returns an
+   * array of the next chunk of previous activity data of the given roomID.
+   *
+   * The next chunk is based on the adapter's `chunkSize`.
+   *
+   * @param {string} ID  ID of the room for which to get activities.
+   * @returns {Array.<string|ActivityDate>}
    * @memberof RoomsAdapter
    */
   // eslint-disable-next-line no-unused-vars
```
[the remark this answers is not in the document]
Comment (lalli-flores) [reply src/adapters/RoomsAdapter.js]: I like that

File: src/adapters/RoomsAdapter.test.js [hunk 8]
```diff
@@ -11,8 +11,8 @@ describe('Rooms Adapter Interface', () => {
     expect(rxjs.isObservable(roomsAdapter.getRoom())).toBeTruthy();
   });
 
-  test('getRoom() throws a proper error message', (done) => {
-    roomsAdapter.getRoom('id').subscribe(
+  test('getRoom() throws error since it is not defined', (done) => {
```
[the remark this answers is not in the document]
Comment (lalli-flores) [reply src/adapters/RoomsAdapter.test.js]: Updated to `because it needs to be defined`. I think it's more explicit why.

File: src/adapters/RoomsAdapter.test.js [hunk 9]
```diff
@@ -21,25 +21,11 @@ describe('Rooms Adapter Interface', () => {
...
-      }
-    );
-  });
-
   test('getRoomActivities() returns an observable', () => {
     expect(rxjs.isObservable(roomsAdapter.getRoomActivities())).toBeTruthy();
   });
 
-  test('getRoomActivities() throws a proper error message', (done) => {
+  test('getRoomActivities() throws error since it is not defined', (done) => {
```
[the remark this answers is not in the document]
Comment (lalli-flores) [reply src/adapters/RoomsAdapter.test.js]: Updated to `because it needs to be defined`. I think it's more explicit why.

File: .storybook/setupTests.js (hunk 1 above)
Comment (lalli-flores) [reply .storybook/setupTests.js]: They are [polyfills](https://babeljs.io/docs/en/next/babel-polyfill.html) so we can use `async`/`await`.
I changed implementation to observable, so this will not be necessary anymore.

File: src/adapters/RoomsJSONAdapter.js [hunk 10]
```diff
@@ -22,9 +22,8 @@ import RoomsAdapter from './RoomsAdapter';
 export default class RoomsJSONAdapter extends RoomsAdapter {
   constructor(datasource) {
     super(datasource);
-    this.getRoom = this.getRoom.bind(this);
-    this.getPreviousRoomActivities = this.getPreviousRoomActivities.bind(this);
-    this.getRoomActivities = this.getRoomActivities.bind(this);
+
```
[the remark this answers is not in the document]
Comment (lalli-flores) [reply src/adapters/RoomsJSONAdapter.js]: Added bindings again

File: src/adapters/RoomsAdapter.test.js [hunk 11]
```diff
@@ -49,6 +35,21 @@ describe('Rooms Adapter Interface', () => {
...
+  test('getPreviousRoomActivities() returns a promise rejection since it is not defined', async () => {
+    try {
+      await roomsAdapter.getPreviousRoomActivities('ID');
+    } catch (error) {
+      expect(error.message).toEqual('getPreviousRoomActivities(ID) must be defined in RoomsAdapter');
+    }
+  });
+
+  test('hasMoreActivities() throws error since it is not defined', () => {
+    try {
```
Comment (adamweeks) [inline src/adapters/RoomsAdapter.test.js]: Does this need to be in a try/catch?

File: src/components/WebexActivityStream/WebexActivityStream.scss (hunk 6 above)
Comment (lalli-flores) [reply src/components/WebexActivityStream/WebexActivityStream.scss]: I tried that, doesn't seem to work.

File: src/components/WebexActivityStream/WebexActivityStream.js [hunk 12]
```diff
@@ -194,8 +197,10 @@ export default function WebexActivityStream({roomID}) {
   });
 
   return (
-    <div className="activity-stream">
+    <div className="activity-stream" ref={activityStreamRef}>
+      {showLoader ? <div className="activity-stream-loader" /> : null}
```
Comment (adamweeks) [inline src/components/WebexActivityStream/WebexActivityStream.js]: I usually see it as:
```js
{ showLoader && <div }
```
but this works as well, not a blocker.

File: src/components/hooks/useActivityStream.js (hunk 2 above)
Comment (lalli-flores) [reply src/components/hooks/useActivityStream.js]: This is how we "fake wait". Data loads too fast and you can't see the loader. 
In general, I think it makes the UX a little better anyway, since we indicate to the user that we actually doing something, even when data comes through really fast. It's more for acknowledgment

File: src/components/hooks/useActivityStream.js (hunk 3 above)
Comment (lalli-flores) [reply src/components/hooks/useActivityStream.js]: Logic has changed, no longer relevant.

File: src/components/hooks/useActivityStream.js (hunk 5 above)
Comment (lalli-flores) [reply src/components/hooks/useActivityStream.js]: Because there are two different sets of side-effects:
1) Side effects to execute only once, on load
2) Side effects when there is a change to the activities

File: src/data/activities.json (hunk 4 above)
Comment (lalli-flores) [reply src/data/activities.json]: Totally added it by accident, and will remove it

File: src/components/hooks/useActivityStream.js [hunk 13]
```diff
@@ -1,53 +1,129 @@
...
+
+  // Keep requesting previous activities until activity stream is filled
   useEffect(() => {
-    const subscription = roomsAdapter.getRoomActivities(roomID).subscribe((activities) => {
-      dispatch({type: APPEND_ACTIVITIES, payload: activities});
+    const activityStreamBoundaries = elementRef.current.getBoundingClientRect();
+    const lastActivityBoundaries = lastElementRef.current.getBoundingClientRect();
+    const hasMoreActivities = roomsAdapter.hasMoreActivities(roomID);
+
+    if (lastActivityBoundaries.bottom < activityStreamBoundaries.bottom) {
```
Comment (adamweeks) [inline src/components/hooks/useActivityStream.js]: Is this when it is at the top? I need more comments to understand what is happening here please

File: src/adapters/RoomsAdapter.test.js (hunk 11 above)
Comment (lalli-flores) [reply src/adapters/RoomsAdapter.test.js]: Yes, otherwise it literally errors out the process.

File: src/adapters/RoomsAdapter.test.js [hunk 14]
```diff
@@ -49,6 +35,20 @@ describe('Rooms Adapter Interface', () => {
     );
   });
 
+  test('getPreviousRoomActivities() throws error since it is not defined', () => {
```
[the remark this answers is not in the document]
Comment (lalli-flores) [reply src/adapters/RoomsAdapter.test.js]: No longer applicable with observables.

File: src/components/hooks/useActivityStream.js (hunk 13 above)
Comment (lalli-flores) [reply src/components/hooks/useActivityStream.js]: Added comment for clarity, let me know if that helps explain

File: src/components/hooks/usePreviousActivities.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  const [lastActivityBottom, setLastActivityBottom] = useState(0);
+
+  useEffect(() => {
+    const activityStreamBoundaries = elementRef.current.getBoundingClientRect();
+    const lastActivityBoundaries = lastElementRef.current.getBoundingClientRect();
+
+    // If the bottom position of the last activity is less than
+    // the bottom position of the container that means that the last
+    // activity is visible and there is not enough content to overflow
+    // the container
```
Comment (adamweeks) [inline src/components/hooks/usePreviousActivities.js]: Does this mean that if the window is resized to be bigger it will automatically load more activities?

File: src/components/WebexActivityStream/WebexActivityStream.js [hunk 16]
```diff
@@ -179,8 +180,16 @@ TimeRuler.propTypes = {
 };
 
 export default function WebexActivityStream({roomID}) {
+  const [activitiesData, dispatch] = useActivityStream(roomID);
+  const loadPreviousActivities = (previousActivities) => {
```
Comment (akoushke) [inline src/components/WebexActivityStream/WebexActivityStream.js]: why is this here instead of being inside the hook?

Comment (lalli-flores) [reply src/components/WebexActivityStream/WebexActivityStream.js]: Because two different hooks need the same function

File: src/components/hooks/usePreviousActivities.js (hunk 15 above)
Comment (lalli-flores) [reply src/components/hooks/usePreviousActivities.js]: It should adjust how much content it loads based on the size of the window, but once it has scrolled to the last activity - meaning it ran once - it will not load content again.

File: src/adapters/RoomsAdapter.js [hunk 17]
```diff
@@ -17,56 +33,55 @@ export const RoomType = {
...
   }
 
   /**
-   * Returns an observable that emits an array of current activities of the given roomID.
+   * Returns `true` if there are more activities to load from the room of the given ID.
+   * Otherwise, it returns `false`.
    *
-   * @param {string} ID  ID of the room to get.
-   * @returns {Observable.<Array.<string|ActivityData>>}
+   * @param {string} ID ID of the room for which to verify activities.
```
Comment (adamweeks) [inline src/adapters/RoomsAdapter.js]: Missing the `@returns` statement

File: src/components/WebexActivityStream/WebexActivityStream.js [hunk 18]
```diff
@@ -179,8 +180,16 @@ TimeRuler.propTypes = {
...
+    dispatch({type: PREPEND_ACTIVITIES, payload: previousActivities});
+  };
+
   const {title, roomType} = useRoom(roomID);
-  const activitiesData = useActivityStream(roomID);
+  const activityStreamRef = useRef(null);
```
Comment (adamweeks) [inline src/components/WebexActivityStream/WebexActivityStream.js]: 👍

File: src/components/hooks/useActivityScroll.js [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+  const {roomsAdapter} = useContext(AdapterContext);
+  const [showLoader, setShowLoader] = useState(false);
+
+  useEffect(() => {
+    let scrollListener = empty();
+
+    if (elementRef.current) {
+      // Listen to scroll event, and on reaching the top,
+      // And load previous activities if there are any
+      scrollListener = fromEvent(elementRef.current, 'scroll')
```
Comment (adamweeks) [inline src/components/hooks/useActivityScroll.js]: This is an awesome use of rxjs!

File: src/components/hooks/useActivityStream.js [hunk 20]
```diff
@@ -2,52 +2,57 @@ import {useContext, useEffect, useReducer} from 'react';
...
+ * @returns {Array.<string|ActivityDate>}
  */
-function reducer(activityIDs, action) {
-  let newActivityIDs = [];
+function reducer(activities, action) {
+  let newActivities = [];
 
   switch (action.type) {
+    case PREPEND_ACTIVITIES:
+      newActivities = action.payload.reverse().concat(activities);
```
Comment (adamweeks) [inline src/components/hooks/useActivityStream.js]: Why do we need to reverse the order of the previous activities? Should the adapter return them in proper order?

File: src/components/hooks/usePreviousActivities.js [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+ * the given callback with that data.
+ * It completes once there is enough data to fill the viewport of the
+ * parent container, finishing by scrolling into view the last activity.
+ *
+ * @param {string} roomID  ID of the room for which to load data.
+ * @param {object} elementRef  DOM reference of the parent component.
+ * @param {previousActivitiesCallback} callback  Callback to execute once data has been fetched.
+ * @returns {object}  DOM reference to the the last activity.
+ */
+export default function usePreviousActivities(roomID, elementRef, callback) {
```
Comment (adamweeks) [inline src/components/hooks/usePreviousActivities.js]: If this hook wasn't defined, wouldn't the top scroll hook catch this case because the scroll top would be the viewport top?

File: src/components/hooks/useActivityScroll.js [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+        .pipe(
+          filter((event) => {
+            const isViewportTop = event.target.scrollTop === 0;
+            const loadMoreItems = isViewportTop && roomsAdapter.hasMoreActivities(roomID);
```
Comment (adamweeks) [inline src/components/hooks/useActivityScroll.js]: When implementing the `hasMoreActivities` in the SDK adapter, we need to be sure to consider that there is no real way to know the room has or doesn't have more activities. The workaround would be that if the amount returned is less than the limit requested.

File: src/adapters/RoomsAdapter.js (hunk 17 above)
Comment (lalli-flores) [reply src/adapters/RoomsAdapter.js]: There was an eslint error because the method doesn't actually return anything. I just disabled the warning for the file and added the return statement.

File: src/components/hooks/usePreviousActivities.js (hunk 21 above)
Comment (lalli-flores) [reply src/components/hooks/usePreviousActivities.js]: It would not, there are 2 different but related use cases.
`useActivityScroll` listens to scroll events, and while there is more data to load, it will load more of the previous activities as you're scrolling up.
`usePreviousActivities` kinda works as an initializer, which loads as much data as necessary to try to overflow the container - based on its size - and then leaves you on the most recent activity. Without `usePreviousActivities`, `useActivityScroll` wouldn't work, because if there is no content loaded or not enough content loaded, the activity stream container will not overflow, and therefore won't have a scroll bar to trigger the scroll event (load more -even older - data). Now, if there is not enough previous activities to overflow the container to begin with, there's no need to listen for scroll events because, well, there's nothing more to load on scroll up.
Before i had all this in `useActivityStream` (combined with the subscription to future events), but i thought separating them out would be more maintainable.

File: src/components/hooks/useActivityScroll.js (hunk 22 above)
Comment (lalli-flores) [reply src/components/hooks/useActivityScroll.js]: really? From the SDK, there's no way to know that there is no more activities to load?

File: src/components/hooks/useActivityStream.js (hunk 20 above)
Comment (lalli-flores) [reply src/components/hooks/useActivityStream.js]: My assumption is that most data retrieved is sorted most-recent to older. Also, displaying the data in reverse order is a function of the UX of the activity stream, and not so much of the data itself, and so, part of the UI logic.
But maybe that's not the case on how data gets sent from the backend?

File: src/components/hooks/usePreviousActivities.js (hunk 21 above)
Comment (adamweeks) [reply src/components/hooks/usePreviousActivities.js]: I like `useOverflowActivities` a lot. It is less confusing to me since `usePreviousActivities` sounds like it is a hook specifically for `getPreviousRoomActivities`.

File: src/components/hooks/useActivityStream.js (hunk 20 above)
Comment (adamweeks) [reply src/components/hooks/useActivityStream.js]: Just investigated and saw that the `/messages` endpoint returns data in reverse date sort (latest to oldest). Conversation endpoint does the opposite (oldest to latest). If we aren't going to do any kind of "sorting" in this hook, we need to make sure to document the expected order. 

We should expect the correct order from the adapter and reduce the code here to just prepend the array to the existing array.

File: src/components/hooks/useActivityScroll.js (hunk 22 above)
Comment (adamweeks) [reply src/components/hooks/useActivityScroll.js]: Correct, take a look at the messages endpoint (https://developer.webex.com/docs/api/v1/messages/list-messages), all we get is an array and no meta-data about remaining messages 😢

File: src/components/hooks/usePreviousActivities.js (hunk 21 above)
Comment (lalli-flores) [reply src/components/hooks/usePreviousActivities.js]: Renamed to `useOverflowActivities`.

File: src/components/hooks/useActivityStream.js (hunk 20 above)
Comment (lalli-flores) [reply src/components/hooks/useActivityStream.js]: My initial instinct is also to have activities sorted newest-to-oldest, as we on the `/messages` API; second reason being consistent with public APIs. However, I've changed and documented the expected order as you requested.

File: src/components/hooks/useActivityScroll.js (hunk 22 above)
Comment (lalli-flores) [reply src/components/hooks/useActivityScroll.js]: Oh that's fine, I think we should be able to know by not getting any more data (like receiving an empty array?). I guess we'll find out on the SDK adapter implementation.
